### PR TITLE
tree2: Implicit field schema

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

SchemaBuilder now accepts `ImplicitFieldSchema` in many places which used to require `FieldSchema`.
This allows `Required` fields to be implicitly specified from just their AllowedTypes.
Additionally in these cases the AllowedTypes can be implicitly specified from a single `Any` or `TreeSchema`.

This change also updates a few arbitrarily selected schema to take advantage of the new syntax.

This change makes no attempt to update support for recursive types to support this. Doing so is likely impossible, and at the very least would require either less safe APIs (ex: no extends cluses) and more complex logic (need to handle the lazy/function cases when normalizing)

## Breaking Changes

For unclear reasons, some seemingly random usages of the SchemaAware API were impacted. This mostly seems to be stuff that was unsound before no longer building, but its unclear why and if anything else was impacted.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree/mockData.ts [hunk 1]
```diff
@@ -154,14 +158,18 @@ export type Person = EditableTree &
 		"editable-tree.Test:Person-1.0.0"
 	>;
 
-export const personData: ContextuallyTypedNodeDataObject = {
+export const personData: SchemaAware.TypedField<
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/test/feature-libraries/editable-tree/mockData.ts:161]: for some reason, users of this which used it as initial tree content with schematize were able to compile despite it lacking strong types. This somehow broke, likley due to the changes to how toDocumentSchema was paramaterized.

This code has been updated to use more specific types as a fix, but its not clear exactly why this change was needed of how many other possible usages would be impacted.

File: experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts [hunk 2]
```diff
@@ -3,18 +3,23 @@
...
+import { FieldKind } from "./modular-schema";
 
 // TODO: tests and examples for this file
 
 /**
  * Builds schema libraries, and the schema within them.
  *
  * @remarks
+ * Field, when inferred from {@link ImplicitFieldSchema}, default to the `Required` {@link FieldKind}.
+ *
```
Comment (Josmithr) [inline experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts:21]:
```suggestion
```

File: experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts [hunk 3]
```diff
@@ -3,18 +3,23 @@
...
+import { AllowedTypes, TreeSchema, FieldSchema, Any, TypedSchemaCollection } from "./typed-schema";
+import { FieldKind } from "./modular-schema";
 
 // TODO: tests and examples for this file
 
 /**
  * Builds schema libraries, and the schema within them.
  *
  * @remarks
+ * Field, when inferred from {@link ImplicitFieldSchema}, default to the `Required` {@link FieldKind}.
```
Comment (Josmithr) [inline experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts:20]: Nit:
```suggestion
 * Fields, when inferred from {@link ImplicitFieldSchema}, default to the `Required` {@link FieldKind}.
```

File: experimental/dds/tree2/src/util/utils.ts [hunk 4]
```diff
@@ -373,6 +373,33 @@ export function objectToMap<MapKey extends string | number | symbol, MapValue>(
...
 }
 
+/**
+ * Convert an object used as a map into a new object used like a map..
+ *
+ * This function must only be used with objects specifically intended to encode map like information.
+ * The only time such objects should be used is for encoding maps as object literals to allow for developer ergonomics or JSON compatibility.
+ * Even those two use-cases need to be carefully considered as using objects as maps can have a lot of issues
+ * (including but not limited to unintended access to __proto__ and other non-owned keys).
+ * {@link objectToMap} helps these few cases get into using an actual map in as safe of was as is practical.
```
Comment (noencke) [inline experimental/dds/tree2/src/util/utils.ts:383]:
```suggestion
 * {@link objectToMap} helps these few cases get into using an actual map in as safe of a way as is practical.
```

File: experimental/dds/tree2/src/util/utils.ts [hunk 5]
```diff
@@ -373,6 +373,33 @@ export function objectToMap<MapKey extends string | number | symbol, MapValue>(
...
+export function transformObjectMap<MapKey extends string | number | symbol, MapValue, NewMapValue>(
+	objectMap: Record<MapKey, MapValue>,
+	transformer: (value: MapValue, key: MapKey) => NewMapValue,
+): Record<MapKey, MapValue> {
+	const output: Record<MapKey, MapValue> = Object.create(null);
+	// This function must only be used with objects specifically intended to encode map like information.
+	for (const key of Object.keys(objectMap)) {
+		const element = objectMap[key as MapKey];
+		Object.defineProperty(output, key, {
+			enumerable: true,
```
Comment (noencke) [inline experimental/dds/tree2/src/util/utils.ts:394]: Perhaps it makes sense to forward the enumerable, configurable, and writable flags from the original property? For example, if the input map's entry is not writeable, then it makes sense that the associated output entry should not be either. I guess enumerable will always be true here since you find the keys by enumerating.

File: experimental/dds/tree2/src/feature-libraries/schema-aware/schemaAware.ts [hunk 6]
```diff
@@ -146,7 +146,7 @@ export type TypedFields<
 > = [
 	TFields extends { [key: string]: FieldSchema }
 		? {
-				[key in keyof TFields]: TypedField<
+				-readonly [key in keyof TFields]: TypedField<
```
Comment (noencke) [inline experimental/dds/tree2/src/feature-libraries/schema-aware/schemaAware.ts:149]: What is the purpose of this?

**Review by noencke — APPROVED** [review]

File: experimental/dds/tree2/src/feature-libraries/schema-aware/schemaAware.ts (hunk 6 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/feature-libraries/schema-aware/schemaAware.ts:149]: It makes the bubble bench demo still compile. Without it, its properties lose their mutability. Changes in struct schema inference make the struct schema tend to be readonly now, and this type used to blindly preserve that when generating the API.

Comment (noencke) [reply experimental/dds/tree2/src/feature-libraries/schema-aware/schemaAware.ts:149]: I see, thanks!

File: experimental/dds/tree2/src/util/utils.ts (hunk 5 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/util/utils.ts:394]: Since this creates a new object which the caller will have the only reference to, it makes sense for it to be mutable. If mutability isn't desired the caller can simply use TypeScript to restrict that (the reverse isn't valid, so this way is more general). Note that the TypeScript type returned requires mutability, so this should match that.

Comment (noencke) [reply experimental/dds/tree2/src/util/utils.ts:394]: Good point, that makes sense.

File: experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts [hunk 7]
```diff
@@ -51,25 +67,29 @@ export class SchemaBuilder<
 	 * Reduced type safety is a side effect of a workaround for a TypeScript limitation.
 	 *
 	 * See note on RecursiveTreeSchema for details.
+	 *
+	 * TODO: Make this work with ImplicitFieldSchema.
```
Comment (Josmithr) [inline experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts:71]: Nit
```suggestion
	 * @privateRemarks TODO: Make this work with ImplicitFieldSchema.
```

File: experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts [hunk 8]
```diff
@@ -51,25 +67,29 @@ export class SchemaBuilder<
 	 * Reduced type safety is a side effect of a workaround for a TypeScript limitation.
 	 *
 	 * See note on RecursiveTreeSchema for details.
```
Comment (Josmithr) [inline experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts:69]: Should we link to `RecursiveTreeSchema` here, since this method shows up in the public API?

File: experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts (hunk 7 above)
Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts:71]: \+ the below methods

File: experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts (hunk 8 above)
Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts:69]: (same with the other methods)

File: experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts [hunk 9]
```diff
@@ -180,12 +209,90 @@ export class SchemaBuilder<
 	): FieldSchema<typeof FieldKinds.sequence, T> {
 		return new FieldSchema(FieldKinds.sequence, t);
 	}
+
+	/**
+	 * Produce a TypedSchemaCollection which captures the content added to this builder, any additional SchemaLibraries that were added to it and a root field.
+	 * Can be used with schematize to provide schema aware access to document content.
```
Comment (Josmithr) [inline experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts:215]: Nit
```suggestion
	 * Can be used with schematize to provide schema aware access to document content.
	 *
	 * @remarks
```

File: experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts [hunk 10]
```diff
@@ -180,12 +209,90 @@ export class SchemaBuilder<
...
+/**
+ * @alpha Normalizes an {@link ImplicitAllowedTypes} into  {@link AllowedTypes}.
+ */
+export type NormalizeAllowedTypes<TSchema extends ImplicitAllowedTypes> = TSchema extends TreeSchema
+	? readonly [TSchema]
+	: TSchema extends Any
+	? readonly [Any]
+	: TSchema;
+
+export function normalizeField<TSchema extends ImplicitFieldSchema, TDefault extends FieldKind>(
```
Comment (Josmithr) [inline experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts:261]: Docs for these functions?

File: experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts [hunk 11]
```diff
@@ -180,12 +209,90 @@ export class SchemaBuilder<
...
+ */
+export type NormalizeField<
+	TSchema extends ImplicitFieldSchema,
+	TDefault extends FieldKind,
+> = TSchema extends FieldSchema
+	? TSchema
+	: FieldSchema<TDefault, NormalizeAllowedTypes<Assume<TSchema, ImplicitAllowedTypes>>>;
+
+/**
+ * @alpha Normalizes an {@link ImplicitAllowedTypes} into  {@link AllowedTypes}.
```
Comment (Josmithr) [inline experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts:253]:
```suggestion
 * Normalizes an {@link ImplicitAllowedTypes} into  {@link AllowedTypes}.
 * @alpha
```

File: experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts [hunk 12]
```diff
@@ -180,12 +209,90 @@ export class SchemaBuilder<
...
 /**
  * Extends {@link SchemaBuilder1} with functionality only used to create built in special libraries.
  * @privateRemarks Should not be package exported.
  */
 export class SchemaBuilderInternal<
 	TScope extends `com.fluidframework.${string}`,
 > extends SchemaBuilder<TScope> {}
+
+/**
+ * @alpha Normalizes an {@link ImplicitFieldSchema} into a {@link FieldSchema}.
```
Comment (Josmithr) [inline experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts:243]:
```suggestion
 * Normalizes an {@link ImplicitFieldSchema} into a {@link FieldSchema}.
 * @alpha
```

File: experimental/dds/tree2/src/test/feature-libraries/schema-aware/schemaSimple.ts [hunk 13]
```diff
@@ -11,8 +11,8 @@ const builder = new SchemaBuilder({ scope: "Simple Schema" });
 export const numberSchema = builder.leaf("number", ValueSchema.Number);
 
 export const pointSchema = builder.struct("point", {
-	x: SchemaBuilder.fieldRequired(numberSchema),
-	y: SchemaBuilder.fieldRequired(numberSchema),
+	x: numberSchema,
```
Comment (Josmithr) [inline experimental/dds/tree2/src/test/feature-libraries/schema-aware/schemaSimple.ts:14]: This is a great improvement!

File: experimental/dds/tree2/src/test/util/utils.spec.ts [hunk 14]
```diff
@@ -22,4 +22,11 @@ describe("Utils", () => {
 			assert.equal(capitalize(input), expected);
 		}
 	});
+
+	it("transformObjectMap", () => {
+		assert.deepEqual(
+			transformObjectMap({ a: "b", c: "d" }, (value, key) => `${key}${value}`),
+			Object.assign(Object.create(null), { a: "ab", c: "cd" }),
```
Comment (Josmithr) [inline experimental/dds/tree2/src/test/util/utils.spec.ts:29]: Just curious, why is the assignment pattern required here?

File: experimental/dds/tree2/src/util/utils.ts [hunk 15]
```diff
@@ -373,6 +373,33 @@ export function objectToMap<MapKey extends string | number | symbol, MapValue>(
 	return map;
 }
 
+/**
+ * Convert an object used as a map into a new object used like a map..
+ *
```
Comment (Josmithr) [inline experimental/dds/tree2/src/util/utils.ts:378]: Nit
```suggestion
 *
 * @remarks
```

**Review by Josmithr — APPROVED** [review]
I left some suggestions, mostly around documentation. Otherwise looks good to me! The builder changes look like a great usability improvement!

File: experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts (hunk 7 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts:71]: This TODO is something callers of this method probably want to know about (specifically that it hasn't been done, and that it might get done). This is doubly true since the callers don't get type safety for this since we can't have extends clauses. Therefor making this part of the API docs was intentional.

Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts:71]: Regular `@remarks` then?

File: experimental/dds/tree2/src/test/util/utils.spec.ts (hunk 14 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/test/util/utils.spec.ts:29]: deepEquals compares object tags and prototypes. This is the cleanest fully spec compliant way I found to make an object with a null prototype for comparison. Putting __proto__ in the object literal apparently works, but only exists for legacy JS compat and is not truly standard. The other option of using object.create and specifiying all the members in there is real ugly as they all have to be property descriptors.

File: experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts (hunk 8 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts]: done

File: experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts (hunk 7 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/feature-libraries/schemaBuilder.ts:71]: I think making it very visible is important since if you don't know this and use this API, you can get really nasty errors elsewhere in your program (like when you try and use the schema), so I'm erroring on the side of making it part of the top level description. Once we determine if this is even possible to fix and/or settle on a different pattern for recursive types, we can convert this situation into more clear docs.
